Merging `tasks.*.json` to `tasks.json` only merges `tasks.local.json`

As the title of this issue for some reason the merging of tasks only merges the local.json file and not both.

### Versions
#### Plugin
```text
Name: Workspace Config+
Id: Swellaby.workspace-config-plus
Description: VS Code extension for improved workspace configuration
Version: 0.2.4
Publisher: Swellaby
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=swellaby.workspace-config-plus
```
#### VSCode
```text
Version: 1.75.1 (user setup)
Commit: 441438abd1ac652551dbe4d408dfcec8a499b8bf
Date: 2023-02-08T21:32:34.589Z
Electron: 19.1.9
Chromium: 102.0.5005.194
Node.js: 16.14.2
V8: 10.2.154.23-electron.0
OS: Windows_NT x64 10.0.19044
Sandboxed: No
```

### Files
#### tasks.shared.json
```json
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "server-sonar",
            "type": "shell",
            "command": "${workspaceFolder}/deployment/podman/run_sonarqube.sh",
            "problemMatcher": []
        },
        {
            "label": "scan-sonar",
            "type": "shell",
            "command": "${workspaceFolder}/deployment/podman/run_scan_sonar.sh",
            "problemMatcher": []
        },
        {
            "label": "test-pipeline",
            "type": "shell",
            "command": "${workspaceFolder}/deployment/podman/run_all_test_pipeline.sh",
            "problemMatcher": []
        },
        {
            "label": "test-all",
            "type": "shell",
            "command": "${workspaceFolder}/deployment/podman/run_all_test.sh",
            "problemMatcher": []
        },
        {
            "label": "test-maven",
            "type": "shell",
            "command": "mvn clean test",
            "problemMatcher": []
        },
        {
            "label": "lint-java",
            "type": "shell",
            "command": "npx prettier --check \"src/**/*.java\"",
            "problemMatcher": []
        }
    ]
}
```

#### tasks.local.json
```json
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "merge-develop",
            "type": "shell",
            "command": "git merge origin/develop",
            "problemMatcher": []
        },
        {
            "label": "lint-all",
            "type": "shell",
            "command": "npx prettier --check .",
            "problemMatcher": []
        }
    ]
}
```
### tasks.json
```
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "merge-develop",
      "type": "shell",
      "command": "git merge origin/develop",
      "problemMatcher": []
    },
    {
      "label": "lint-all",
      "type": "shell",
      "command": "npx prettier --check .",
      "problemMatcher": []
    }
  ]
}
```

### Tests
I update `tasks.local.json` with in the array of tasks
```json
    {
      "label": "pull",
      "type": "shell",
      "command": "git pull",
      "problemMatcher": []
    }
```
which produceses on save `[2023-03-01 09:10:40] [INFO] Updating config in /home/eriase/repos/quarkus-maven-project/.vscode/tasks.json`.

After this I remove the added line from `tasks.local.json` the same log is produced and add it to update the tasks.shared.json and no log is produced at all ☹

If I add an incomplete json object like below the it failes to parse the `tasks.shared.json` file as excepted.
```json
       {
            "label":
        }
```
```log
[2023-03-01 09:18:54] [ERROR] Failed to parse contents of: /home/eriase/repos/quarkus-maven-project/.vscode/tasks.shared.json
[2023-03-01 09:18:54] [DEBUG] Error: Failed to parse contents of: /home/eriase/repos/quarkus-maven-project/.vscode/tasks.shared.json
```

--- a/package.json
+++ b/package.json
@@ -55,7 +55,24 @@
         "fileMatch": "/.vscode/tasks.local.json",
         "url": "vscode://schemas/tasks"
       }
-    ]
+    ],
+    "configuration": {
+      "type": "object",
+      "title": "Workspace Config+",
+      "properties": {
+        "workspaceConfigPlus.arrayMerge": {
+          "description": "Defines the merge behavior to use for arrays",
+          "type": "string",
+          "scope": "resource",
+          "default": "combine",
+          "enum": ["combine", "overwrite"],
+          "enumDescriptions": [
+            "Performs a deep merge of array-type properties overlapping in both local and shared user settings, where the arrays are combined",
+            "If both the local and shared files define a value for the same array-type property, then the array value in the local file is used and the shared array value is ignored"
+          ]
+        }
+      }
+    }
   },
   "keywords": [
     "local settings",

--- a/src/file-handler.js
+++ b/src/file-handler.js
@@ -5,6 +5,9 @@ const jsoncParser = require('jsonc-parser');
 const deepMerge = require('deepmerge');
 const log = require('./log');
 
+const _arrayMergeKey = 'workspaceConfigPlus.arrayMerge';
+const _arrayMergeDefaultValue = 'combine';
+
 const _loadConfigFromFile = async (fileUri, readFile) => {
   const contents = await readFile(fileUri);
   if (!contents) {
@@ -20,6 +23,26 @@ const _loadConfigFromFile = async (fileUri, readFile) => {
   return config;
 };
 
+const getMergedConfigs = ({ sharedConfig, localConfig }) => {
+  const shared = sharedConfig || {};
+  const local = localConfig || {};
+  let arrayMerge =
+    local[_arrayMergeKey] || shared[_arrayMergeKey] || _arrayMergeDefaultValue;
+  const invalidValueErrorMessage = `Invalid value for 'arrayMerge' setting: '${arrayMerge}'. Must be 'overwrite' or 'combine'`;
+  if (typeof arrayMerge != 'string') {
+    throw new Error(invalidValueErrorMessage);
+  }
+
+  arrayMerge = arrayMerge.toLowerCase();
+  let options = {};
+  if (arrayMerge == 'overwrite') {
+    options.arrayMerge = (_dest, source, _options) => source;
+  } else if (arrayMerge !== 'combine') {
+    throw new Error(invalidValueErrorMessage);
+  }
+  return deepMerge(shared, local, options);
+};
+
 // This function does exceed our preferred ceiling for statement counts
 // but worth an override here for readability. However, we should split
 // this up if we end up needing to add anything else to it.
@@ -33,21 +56,19 @@ const mergeConfigFiles = async ({
 }) => {
   const loadConfigFromFile = module.exports._loadConfigFromFile;
   try {
-    const sharedFile = await loadConfigFromFile(sharedFileUri, readFile);
-    const localFile = await loadConfigFromFile(localFileUri, readFile);
+    const sharedConfig = await loadConfigFromFile(sharedFileUri, readFile);
+    const localConfig = await loadConfigFromFile(localFileUri, readFile);
 
     // If neither of these files exists then there's no work to be done
-    if (!sharedFile && !localFile) {
+    if (!sharedConfig && !localConfig) {
       return;
     }
 
-    const sharedFileContents = sharedFile || {};
-    const localFileContents = localFile || {};
     const vscodeFileContents = await loadConfigFromFile(
       vscodeFileUri,
       readFile
     );
-    const merged = deepMerge(sharedFileContents, localFileContents);
+    const merged = getMergedConfigs({ sharedConfig, localConfig });
 
     // Avoid rewriting the file if there are no changes to be applied
     if (isDeepStrictEqual(vscodeFileContents, merged)) {
@@ -58,7 +79,7 @@ const mergeConfigFiles = async ({
     await writeFile(
       vscodeFileUri,
       Buffer.from(
-        JSON.stringify(deepMerge(vscodeFileContents, merged), null, 2)
+        JSON.stringify({ ...vscodeFileContents, ...merged }, null, 2)
       ),
       { create: true, overwrite: true }
     );
@@ -72,4 +93,6 @@ module.exports = {
   mergeConfigFiles,
   // Private, only exported for test mocking
   _loadConfigFromFile,
+  _arrayMergeKey,
+  _arrayMergeDefaultValue,
 };

--- a/tests/unit/config.js
+++ b/tests/unit/config.js
@@ -1,0 +1,24 @@
+'use strict';
+
+const { assert } = require('chai');
+
+const fileHandler = require('../../src/file-handler');
+const {
+  contributes: { configuration },
+} = require('../../package.json');
+const { properties: config } = configuration;
+
+suite('config Suite', () => {
+  test('Should have the correct number of configuration settings', () => {
+    assert.deepEqual(Object.keys(config).length, 1);
+  });
+
+  test('Should have correct config setting values for array merge behavior', () => {
+    const arrayMergeConfig = config[fileHandler._arrayMergeKey];
+    assert.deepEqual(
+      arrayMergeConfig.default,
+      fileHandler._arrayMergeDefaultValue
+    );
+    assert.deepEqual(arrayMergeConfig.enum, ['combine', 'overwrite']);
+  });
+});

--- a/tests/unit/file-handler.js
+++ b/tests/unit/file-handler.js
@@ -2,11 +2,10 @@
 
 const { assert } = require('chai');
 const jsoncParser = require('jsonc-parser');
-const deepMerge = require('deepmerge');
 const Sinon = require('sinon');
 
 const { callbacks } = require('../data');
-const fileHandler = require('../..//src/file-handler');
+const fileHandler = require('../../src/file-handler');
 const log = require('../../src/log');
 
 suite('file handler Suite', () => {
@@ -73,6 +72,8 @@ suite('file handler Suite', () => {
     /** @type {Sinon.SinonStub} */
     let loadConfigFromFileStub;
     /** @type {Sinon.SinonStub} */
+    let loadVSConfigFromFileStub;
+    /** @type {Sinon.SinonStub} */
     let writeFileStub;
     /** @type {Sinon.SinonStub} */
     let logInfoStub;
@@ -84,15 +85,21 @@ suite('file handler Suite', () => {
     const sharedFileUri = { path: '.vscode/settings.shared.json' };
     const localFileUri = { path: '.vscode/settings.local.json' };
     const mergeConfigFiles = fileHandler.mergeConfigFiles;
-    const sharedConfig = {
+    const vscodeConfig = {
+      foo: 'abc',
+      'window.zoomLevel': 0,
+      bar: 'def',
+    };
+    const sharedArrayCombineConfig = {
       foo: false,
       'editor.rulers': [100],
       '[typescript]': {
         'editor.dragAndDrop': false,
         'editor.tabCompletion': 'on',
       },
+      [fileHandler._arrayMergeKey]: 'combine',
     };
-    const localConfig = {
+    const localArrayCombineConfig = {
       foo: true,
       'window.zoomLevel': 1,
       'editor.rulers': [80],
@@ -101,36 +108,88 @@ suite('file handler Suite', () => {
         'editor.autoIndent': false,
       },
       baz: false,
+      [fileHandler._arrayMergeKey]: 'combine',
     };
-    const vscodeConfig = {
-      foo: 'abc',
-      'window.zoomLevel': 0,
-      bar: 'def',
-    };
-    const expConfig = {
+
+    const expArrayCombineConfig = {
       foo: true,
       'window.zoomLevel': 1,
-      bar: 'def',
       'editor.rulers': [100, 80],
       '[typescript]': {
         'editor.dragAndDrop': true,
         'editor.tabCompletion': 'on',
         'editor.autoIndent': false,
       },
+      [fileHandler._arrayMergeKey]: 'combine',
+      baz: false,
+    };
+    const finalExpArrayCombineConfig = {
+      ...vscodeConfig,
+      ...expArrayCombineConfig,
+    };
+    const expArrayCombineConfigNoExplicitMerge = JSON.parse(
+      JSON.stringify(finalExpArrayCombineConfig)
+    );
+    delete expArrayCombineConfigNoExplicitMerge[fileHandler._arrayMergeKey];
+
+    const sharedArrayOverwriteConfig = {
+      foo: false,
+      'editor.rulers': [100],
+      '[typescript]': {
+        'editor.dragAndDrop': false,
+        'editor.tabCompletion': 'on',
+      },
+      // Note the casing in the value, added intentionally to provide
+      // defensive testing and ensure we aren't pedantic about case.
+      [fileHandler._arrayMergeKey]: 'OVERwrite',
+    };
+    const sharedConfigNoArrayMerge = JSON.parse(
+      JSON.stringify(sharedArrayOverwriteConfig)
+    );
+    delete sharedConfigNoArrayMerge[fileHandler._arrayMergeKey];
+    const localArrayOverwriteConfig = {
+      foo: true,
+      'window.zoomLevel': 1,
+      'editor.rulers': [80],
+      '[typescript]': {
+        'editor.dragAndDrop': true,
+        'editor.autoIndent': false,
+      },
+      [fileHandler._arrayMergeKey]: 'overwrite',
+      baz: false,
+    };
+    const localConfigNoArrayMerge = JSON.parse(
+      JSON.stringify(localArrayOverwriteConfig)
+    );
+    delete localConfigNoArrayMerge[fileHandler._arrayMergeKey];
+    const expArrayOverwriteConfig = {
+      foo: true,
+      'window.zoomLevel': 1,
+      bar: 'def',
+      'editor.rulers': [80],
+      '[typescript]': {
+        'editor.dragAndDrop': true,
+        'editor.tabCompletion': 'on',
+        'editor.autoIndent': false,
+      },
+      [fileHandler._arrayMergeKey]: 'overwrite',
       baz: false,
     };
 
     setup(() => {
       loadConfigFromFileStub = Sinon.stub(fileHandler, '_loadConfigFromFile');
+      loadVSConfigFromFileStub = loadConfigFromFileStub.withArgs(
+        vscodeFileUri,
+        callbacks.readFile
+      );
       loadConfigFromFileStub
         .withArgs(sharedFileUri, callbacks.readFile)
-        .callsFake(() => Promise.resolve(sharedConfig));
+        .callsFake(() => Promise.resolve(sharedArrayCombineConfig));
       loadConfigFromFileStub
         .withArgs(localFileUri, callbacks.readFile)
-        .callsFake(() => Promise.resolve(localConfig));
-      loadConfigFromFileStub
-        .withArgs(vscodeFileUri, callbacks.readFile)
-        .callsFake(() => Promise.resolve(vscodeConfig));
+        .callsFake(() => Promise.resolve(localArrayCombineConfig));
+
+      loadVSConfigFromFileStub.callsFake(() => Promise.resolve(vscodeConfig));
       writeFileStub = Sinon.stub(callbacks, 'writeFile');
       logDebugStub = Sinon.stub(log, 'debug');
       logErrorStub = Sinon.stub(log, 'error');
@@ -155,11 +214,9 @@ suite('file handler Suite', () => {
     });
 
     test('Should return early when there are no changes to be made', async () => {
-      loadConfigFromFileStub
-        .withArgs(vscodeFileUri, callbacks.readFile)
-        .callsFake(() => {
-          return Promise.resolve(deepMerge(sharedConfig, localConfig));
-        });
+      loadVSConfigFromFileStub.callsFake(() => {
+        return Promise.resolve(expArrayCombineConfig);
+      });
       await mergeConfigFiles({
         vscodeFileUri,
         sharedFileUri,
@@ -190,7 +247,11 @@ suite('file handler Suite', () => {
         writeFileStub.calledOnceWithExactly(
           vscodeFileUri,
           Buffer.from(
-            JSON.stringify({ ...vscodeConfig, ...sharedConfig }, null, 2)
+            JSON.stringify(
+              { ...vscodeConfig, ...sharedArrayCombineConfig },
+              null,
+              2
+            )
           ),
           { create: true, overwrite: true }
         )
@@ -217,14 +278,18 @@ suite('file handler Suite', () => {
         writeFileStub.calledOnceWithExactly(
           vscodeFileUri,
           Buffer.from(
-            JSON.stringify({ ...vscodeConfig, ...localConfig }, null, 2)
+            JSON.stringify(
+              { ...vscodeConfig, ...localArrayCombineConfig },
+              null,
+              2
+            )
           ),
           { create: true, overwrite: true }
         )
       );
     });
 
-    test('Should write to config file with correct priority order', async () => {
+    test('Should write to config file with correct priority order using array combine', async () => {
       await mergeConfigFiles({
         vscodeFileUri,
         sharedFileUri,
@@ -240,7 +305,141 @@ suite('file handler Suite', () => {
       assert.isTrue(
         writeFileStub.calledOnceWithExactly(
           vscodeFileUri,
-          Buffer.from(JSON.stringify(expConfig, null, 2)),
+          Buffer.from(JSON.stringify(finalExpArrayCombineConfig, null, 2)),
+          { create: true, overwrite: true }
+        )
+      );
+    });
+
+    test('Should write to config file with correct priority order using shared array overwrite', async () => {
+      loadConfigFromFileStub
+        .withArgs(sharedFileUri, callbacks.readFile)
+        .callsFake(() => Promise.resolve(sharedArrayOverwriteConfig));
+      loadConfigFromFileStub
+        .withArgs(localFileUri, callbacks.readFile)
+        .callsFake(() => Promise.resolve(localArrayCombineConfig));
+      await mergeConfigFiles({
+        vscodeFileUri,
+        sharedFileUri,
+        localFileUri,
+        ...callbacks,
+      });
+      assert.isTrue(loadConfigFromFileStub.calledThrice);
+      assert.isTrue(
+        logInfoStub.calledOnceWithExactly(
+          `Updating config in ${vscodeFileUri.fsPath}`
+        )
+      );
+      assert.isTrue(
+        writeFileStub.calledOnceWithExactly(
+          vscodeFileUri,
+          Buffer.from(JSON.stringify(finalExpArrayCombineConfig, null, 2)),
+          { create: true, overwrite: true }
+        )
+      );
+    });
+
+    test('Should write to config file with correct priority order using shared array overwrite', async () => {
+      loadConfigFromFileStub
+        .withArgs(localFileUri, callbacks.readFile)
+        .callsFake(() => Promise.resolve(localArrayOverwriteConfig));
+      await mergeConfigFiles({
+        vscodeFileUri,
+        sharedFileUri,
+        localFileUri,
+        ...callbacks,
+      });
+      assert.isTrue(loadConfigFromFileStub.calledThrice);
+      assert.isTrue(
+        logInfoStub.calledOnceWithExactly(
+          `Updating config in ${vscodeFileUri.fsPath}`
+        )
+      );
+      assert.isTrue(
+        writeFileStub.calledOnceWithExactly(
+          vscodeFileUri,
+          Buffer.from(JSON.stringify(expArrayOverwriteConfig, null, 2)),
+          { create: true, overwrite: true }
+        )
+      );
+    });
+
+    test('Should write to config file with correct priority order using correct array merge default', async () => {
+      loadConfigFromFileStub
+        .withArgs(sharedFileUri, callbacks.readFile)
+        .callsFake(() => Promise.resolve(sharedConfigNoArrayMerge));
+      loadConfigFromFileStub
+        .withArgs(localFileUri, callbacks.readFile)
+        .callsFake(() => Promise.resolve(localConfigNoArrayMerge));
+      await mergeConfigFiles({
+        vscodeFileUri,
+        sharedFileUri,
+        localFileUri,
+        ...callbacks,
+      });
+      assert.isTrue(loadConfigFromFileStub.calledThrice);
+      assert.isTrue(
+        logInfoStub.calledOnceWithExactly(
+          `Updating config in ${vscodeFileUri.fsPath}`
+        )
+      );
+      assert.isTrue(
+        writeFileStub.calledOnceWithExactly(
+          vscodeFileUri,
+          Buffer.from(
+            JSON.stringify(expArrayCombineConfigNoExplicitMerge, null, 2)
+          ),
+          { create: true, overwrite: true }
+        )
+      );
+    });
+
+    test('Should maintain idempotency for unmodified settings', async () => {
+      loadVSConfigFromFileStub.onSecondCall().callsFake(() => {
+        return Promise.resolve(finalExpArrayCombineConfig);
+      });
+      const updatedLocalConfig = JSON.parse(
+        JSON.stringify(localArrayCombineConfig)
+      );
+      updatedLocalConfig.cow = 'moo';
+      const secondExpectedConfig = JSON.parse(
+        JSON.stringify(finalExpArrayCombineConfig)
+      );
+      secondExpectedConfig.cow = 'moo';
+      loadConfigFromFileStub
+        .withArgs(localFileUri, callbacks.readFile)
+        .onSecondCall()
+        .callsFake(() => Promise.resolve(updatedLocalConfig));
+      await mergeConfigFiles({
+        vscodeFileUri,
+        sharedFileUri,
+        localFileUri,
+        ...callbacks,
+      });
+      await mergeConfigFiles({
+        vscodeFileUri,
+        sharedFileUri,
+        localFileUri,
+        ...callbacks,
+      });
+
+      assert.deepEqual(loadConfigFromFileStub.callCount, 6);
+      assert.isTrue(logInfoStub.calledTwice);
+      assert.isTrue(
+        logInfoStub.calledWith(`Updating config in ${vscodeFileUri.fsPath}`)
+      );
+      assert.isTrue(writeFileStub.calledTwice);
+      assert.isTrue(
+        writeFileStub.firstCall.calledWithExactly(
+          vscodeFileUri,
+          Buffer.from(JSON.stringify(finalExpArrayCombineConfig, null, 2)),
+          { create: true, overwrite: true }
+        )
+      );
+      assert.isTrue(
+        writeFileStub.secondCall.calledWithExactly(
+          vscodeFileUri,
+          Buffer.from(JSON.stringify(secondExpectedConfig, null, 2)),
           { create: true, overwrite: true }
         )
       );
@@ -257,6 +456,44 @@ suite('file handler Suite', () => {
       });
       assert.isTrue(logErrorStub.calledOnceWithExactly(err.message));
       assert.isTrue(logDebugStub.calledOnceWithExactly(err));
+    });
+
+    test('Throws correct error on invalid type for array merge behavior', async () => {
+      const arrayMerge = 2;
+      const invalidArrayMergeTypeConfig = JSON.parse(
+        JSON.stringify(localArrayCombineConfig)
+      );
+      invalidArrayMergeTypeConfig[fileHandler._arrayMergeKey] = arrayMerge;
+      loadConfigFromFileStub
+        .withArgs(localFileUri, callbacks.readFile)
+        .callsFake(() => Promise.resolve(invalidArrayMergeTypeConfig));
+      const expErrMessage = `Invalid value for 'arrayMerge' setting: '${arrayMerge}'. Must be 'overwrite' or 'combine'`;
+      await mergeConfigFiles({
+        vscodeFileUri,
+        sharedFileUri,
+        localFileUri,
+        ...callbacks,
+      });
+      assert.isTrue(logErrorStub.calledOnceWithExactly(expErrMessage));
+    });
+
+    test('Throws correct error on invalid value for array merge behavior', async () => {
+      const arrayMerge = 'shuffle';
+      const invalidArrayMergeValueConfig = JSON.parse(
+        JSON.stringify(localArrayCombineConfig)
+      );
+      invalidArrayMergeValueConfig[fileHandler._arrayMergeKey] = arrayMerge;
+      loadConfigFromFileStub
+        .withArgs(localFileUri, callbacks.readFile)
+        .callsFake(() => Promise.resolve(invalidArrayMergeValueConfig));
+      const expErrMessage = `Invalid value for 'arrayMerge' setting: '${arrayMerge}'. Must be 'overwrite' or 'combine'`;
+      await mergeConfigFiles({
+        vscodeFileUri,
+        sharedFileUri,
+        localFileUri,
+        ...callbacks,
+      });
+      assert.isTrue(logErrorStub.calledOnceWithExactly(expErrMessage));
     });
   });
 });

